File: SourceAFIS/Matching/Minutia/PairSelector.cs
using System;
using System.Collections.Generic;
using System.Text;
using SourceAFIS.General;

namespace SourceAFIS.Matching.Minutia
{
    public sealed class PairSelector
    {
        PriorityQueueF<EdgePair> Queue = new PriorityQueueF<EdgePair>();

        public void Clear()
        {
            Queue.Clear();
        }

        public void Enqueue(EdgePair pair, float distance)
        {
            Queue.Enqueue(distance, pair);
        }

        public void SkipPaired(MinutiaPairing pairing)
        {
            while (Queue.Count > 0 && (pairing.IsProbePaired(Queue.Peek().Neighbor.Probe)
                || pairing.IsCandidatePaired(Queue.Peek().Neighbor.Candidate)))
            {
                EdgePair edge = Queue.Dequeue();
                if (pairing.IsProbePaired(edge.Neighbor.Probe) && pairing.GetByProbe(edge.Neighbor.Probe).Pair.Candidate == edge.Neighbor.Candidate)
                    pairing.AddSupportByProbe(edge.Neighbor.Probe);
            }
        }

        public int Count { get { return Queue.Count; } }

        public EdgePair Dequeue()
        {
            return Queue.Dequeue();
        }
    }
}

File: java/sourceafis/src/sourceafis/matching/minutia/PairSelector.java
package sourceafis.matching.minutia;

import sourceafis.general.PriorityQueueF;

public final class PairSelector
{
    PriorityQueueF<EdgePair> queue = new PriorityQueueF<EdgePair>();

    public void clear()
    {
        queue.clear();
    }

    public void enqueue(EdgePair pair, float distance)
    {
        queue.enqueue(distance, pair);
    }

    public void skipPaired(MinutiaPairing pairing)
    {
        while (queue.size() > 0 && (pairing.isProbePaired(queue.peek().neighbor.probe)
            || pairing.isCandidatePaired(queue.peek().neighbor.candidate)))
        {
            EdgePair edge = (EdgePair)queue.dequeue();
            if (pairing.isProbePaired(edge.neighbor.probe) && pairing.getByProbe(edge.neighbor.probe).pair.candidate == edge.neighbor.candidate) {
            	pairing.addSupportByProbe(edge.reference.probe);
                pairing.addSupportByProbe(edge.neighbor.probe);
            }
        }
    }

    public int getCount() {  return queue.size();  }

    public EdgePair dequeue()
    {
        return  queue.dequeue();
    }
}

